### PR TITLE
testapi: Optimize compat_args()'s handling of one fixed parameter

### PR DESCRIPTION
**Pull request by cfconrad**
With only one fixed parameter, the number of arguments specify
if the fixed parameter was given or not.
It makes the check for known named parameter keys obsolete and
allow additional parameters also in the first field (fixed parameter).

### CODE REVIEW

**Comment by Martchus** [comment]
The commit message has grammar/punctuation issues and I'm also not quite sure whether I understand it correctly. "We do not *need*" and "it is *possible* to *just*" makes it sound like this was just and optimization and there's no change in behavior. Is that correct? Then it would make sense to title the commit e.g. "testapi: Optimize `compat_args()`'s handling of one fixed parameter". Otherwise it would make sense to start with e.g. "testapi: Fix …" to state the intention of the change.

File: testapi.pm [hunk 1]
```diff
@@ -2283,8 +2283,14 @@ A typical call would look like:
 sub compat_args {    # no:style:signatures
     my ($def_args, $fix_keys) = splice(@_, 0, 2);
     my %ret;
-    for my $key (@{$fix_keys}) {
-        $ret{$key} = shift if (scalar(@_) >= 1 && (!defined($_[0]) || !grep { $_ eq $_[0] } keys(%{$def_args})));
```
Comment (perlpunk) [inline testapi.pm:2287]: Not related to the PR, but I never looked closely at this code. It occurred to me that for looking if a key exists in a hash, you don't need grep:
```
# old:
!grep { $_ eq $_[0] } keys(%{$def_args}
# new
! exists $def_args->{$_[0]}
```
That would be a simple optimization.

File: testapi.pm [hunk 2]
```diff
@@ -2283,8 +2283,14 @@ A typical call would look like:
...
+    if (scalar(@$fix_keys) == 1) {
```
Comment (perlpunk) [inline testapi.pm:2286]:
```suggestion
    if (@$fix_keys == 1) {
```

File: testapi.pm [hunk 3]
```diff
@@ -2283,8 +2283,14 @@ A typical call would look like:
...
+        if ((scalar(@_) % 2) != 0) {
```
Comment (perlpunk) [inline testapi.pm:2287]:
```suggestion
        if ((@_ % 2) != 0) {
```

File: testapi.pm [hunk 4]
```diff
@@ -2283,8 +2283,14 @@ A typical call would look like:
...
+            $ret{$fix_keys->[0]} = shift(@_);
```
Comment (perlpunk) [inline testapi.pm:2288]:
```suggestion
            $ret{$fix_keys->[0]} = shift;
```

File: testapi.pm (hunk 1 above)
Comment (cfconrad) [reply testapi.pm:2287]: thx, I have added it anyway...

File: testapi.pm (hunk 4 above)
Comment (cfconrad) [reply testapi.pm]: and I changed it to use `signatures` where we do not have this implicit shift...

**Review by kalikiana — APPROVED** [review]

**Review by okurz — APPROVED** [review]

**Pull request merged**